src/Microsoft.ML.Recommender/SafeTrainingAndModelBuffer.cs

```diff
@@ -133,19 +133,19 @@ private struct MFParameter
             /// Specify if the factor matrices should be non-negative.
             /// </summary>
             [FieldOffset(48)]
-            public int DoNmf;
+            public byte DoNmf;
```

Comment:
While this is better than what we had before, this is still not correct.

[https://docs.microsoft.com/en-us/cpp/cpp/fundamental-types-cpp?view=vs-2019](https://docs.microsoft.com/en-us/cpp/cpp/fundamental-types-cpp?view=vs-2019)

>  bool Type bool is an integral type that can have one of the two values true or false. Its size is unspecified.

[https://en.cppreference.com/w/cpp/language/types](https://en.cppreference.com/w/cpp/language/types)

> bool - type, capable of holding one of the two values: true or false. The value of sizeof(bool) is implementation defined and might differ from 1.

Instead, to make interop/marshalling work correctly on all computers - the technique we use in .NET is to ensure the ABI (application binary interface) is stable no matter the compiler/processor/etc. This means only stable-sized types can be used in the ABI that C# code P/Invokes into.

So to fix this, the libmf C code should change to have a stable ABI. I see it uses `float`, `int`, and `bool`. For sure `int` and `bool` can change sizes depending on the compiler/processor. Instead, we should be using `int32_t` and `int8_t` on the C side.  (Note: you can just have a wrapper export function in the C library that exposes the stable ABI and calls into your current code. So the main C/C++ code can remain unchanged.)

Reply:
Thank you for pointing this out. Another iteration is pushed. Hopefully, it can address this comment.